plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/ProjectBuilder.java

```diff
@@ -143,87 +156,170 @@ public void initialize(ProjectDescription description, XIProjectConfig projectCo
-			List<IResourceDescription.Delta> externalDeltas, CancelIndicator cancelIndicator) {
+	public XBuildResult doIncrementalBuild(
+			IBuildRequestFactory buildRequestFactory,
+			Set<URI> dirtyFiles,
+			Set<URI> deletedFiles,
+			List<IResourceDescription.Delta> externalDeltas,
+			CancelIndicator cancelIndicator) {
+
+		return doBuild(createIncrementalBuildFactory(buildRequestFactory), dirtyFiles, deletedFiles, externalDeltas,
+				cancelIndicator);
+	}
 
-		return doBuild(dirtyFiles, deletedFiles, externalDeltas, true, cancelIndicator);
+	/**
+	 * Create a wrapper around the given build request factory that yields a build request which will update the stored
+	 * project state via {@link ProjectBuilder#updateProjectState}.
+	 */
+	protected ProjectStateUpdatingBuildRequestFactory createIncrementalBuildFactory(
+			IBuildRequestFactory buildRequestFactory) {
+		return new ProjectStateUpdatingBuildRequestFactory(buildRequestFactory, new ProjectStateUpdater());
 	}
 
-	/** Build this project. */
-	protected XBuildResult doBuild(Set<URI> dirtyFiles, Set<URI> deletedFiles,
-			List<IResourceDescription.Delta> externalDeltas, boolean propagateIssues, CancelIndicator cancelIndicator) {
+	/** Build this project according to the request provided by the given buildRequestFactory. */
+	protected XBuildResult doBuild(
+			IBuildRequestFactory buildRequestFactory,
+			Set<URI> dirtyFiles,
+			Set<URI> deletedFiles,
+			List<IResourceDescription.Delta> externalDeltas,
+			CancelIndicator cancelIndicator) {
 
-		URI persistenceFile = projectStateManager.getPersistenceFile();
+		URI persistenceFile = getPersistenceFile();
 		dirtyFiles.remove(persistenceFile);
 		deletedFiles.remove(persistenceFile);
 
-		XBuildRequest request = newBuildRequest(dirtyFiles, deletedFiles, externalDeltas, propagateIssues,
+		XBuildRequest request = newBuildRequest(
+				buildRequestFactory,
+				dirtyFiles,
+				deletedFiles,
+				externalDeltas,
 				cancelIndicator);
-		resourceSet = request.getResourceSet(); // resourceSet is already used during the build via #getResource(URI)
-
-		XBuildResult result = incrementalBuilder.build(request);
-
-		projectStateManager.updateProjectState(request, result);
+		resourceSet = request.getResourceSet();
 
-		ResourceDescriptionsData resourceDescriptions = projectStateManager.getIndex();
-		fullIndex.setProjectIndex(projectDescription.getName(), resourceDescriptions);
-
-		return result;
+		return incrementalBuilder.build(request);
 	}
 
 	/** Deletes the contents of the output directory */
-	public void doClean(CancelIndicator cancelIndicator) {
-		projectStateManager.deletePersistenceFile();
+	public void doClean(AfterDeleteListener deleteListener, CancelIndicator cancelIndicator) {
+		deletePersistenceFile();
 
 		if (projectConfig instanceof N4JSProjectConfig) {
 			// TODO: merge N4JSProjectConfig#canClean() to IProjectConfig
+			// TODO: extract to N4JSProjectBuilder sub-type
```

Comment:
Actually, it might be reasonable to let the API this use case, since it might be also relevant for other users of Xtext?

Reply:
As discussed: Remove canBuild and use indexOnly instead.